R/PipeOpFDAInterpol.R

```diff
@@ -0,0 +1,140 @@
+#'   * `"locf"`: applies "last observation carried forward" interpolation (see `tf::tf_approx_locf()`).
+#'   * `"nocb"`: applies "next observation carried backward" interpolation (see `tf::tf_approx_nocb()`).
+#' * `left` :: `numeric()` \cr
+#'   The left boundary of the window.
+#'   The window is specified such that the all values >=left and <=right are kept for the computations.
+#' * `right` :: `numeric()` \cr
+#'   The right boundary of the window.
+#'
+#' @export
+#' @examples
+#' library(mlr3pipelines)
+#' task = tsk("fuel")
+#' pop = po("fda.interpol")
+#' task_interpol = pop$train(list(task))
+PipeOpFDAInterpol = R6Class("PipeOpFDAInterpol",
+  inherit = mlr3pipelines::PipeOpTaskPreprocSimple,
+  public = list(
+    #' @description Initializes a new instance of this Class.
+    #' @param id (`character(1)`)\cr
+    #'   Identifier of resulting object, default `"fda.interpol"`.
+    #' @param param_vals (named `list`)\cr
+    #'   List of hyperparameter settings, overwriting the hyperparameter settings that would
+    #'   otherwise be set during construction. Default `list()`.
+    initialize = function(id = "fda.interpol", param_vals = list()) {
+      param_set = ps(
+        grid = p_uty(tags = c("train", "predict", "required"), custom_check = crate(function(x) {
+          if (test_string(x)) {
+            return(check_choice(x, choices = c("union", "intersect", "minmax")))
+          }
+          if (test_numeric(x, any.missing = FALSE, min.len = 1)) {
+            return(TRUE)
+          }
+          "Must be either a string or numeric vector."
+        })),
+        method = p_fct(
+          c("linear", "spline", "fill_extend", "locf", "nocb"), default = "linear", tags = c("train", "predict")
+        ),
+        left = p_dbl(tags = c("train", "predict")),
+        right = p_dbl(tags = c("train", "predict"))
+      )
+      param_set$set_values(grid = "union")
+
+      super$initialize(
+        id = id,
+        param_set = param_set,
+        param_vals = param_vals,
+        packages = c("mlr3fda", "mlr3pipelines"),
+        feature_types = c("tfd_reg", "tfd_irreg")
+      )
+    }
+  ),
+  private = list(
+    .transform_dt = function(dt, levels) {
+      pars = self$param_set$get_values()
+      grid = pars$grid
+      method = pars$method
+      left = pars$left
+      right = pars$right
+      if (!is.null(left) && !is.null(right)) {
+        assert_count(grid)
+        assert_true(left <= right)
+      }
+      method = method %??% "linear"
+      evaluator = sprintf("tf_approx_%s", method)
+      map_dtc(dt, function(x) interpolate_col(x, grid, evaluator, left, right))
+    }
+  )
+)
+
+interpolate_col = function(x, grid, evaluator, left, right) {
+  if (is.numeric(grid)) {
+    if (length(grid) > 1L && is.null(left) && is.null(right)) {
+      arg = unlist(tf::tf_arg(x))
+      if (max(grid) > max(arg) || min(grid) < min(arg)) {
```

Comment:
this calculates the max(grid) and min(grid) for each column, do this once and pass it as arguments to interpolate_col

Reply:
Why do this once? Then its the same for all cols is this wanted?

Reply:
I mean the grid is the same for all cols anyway, no?

Reply:
No why would it be the same for all cols?

Reply:
because the grid is one parameter

Reply:
I think you are confusing grid with arg, I am only talking about grid

Reply:
Actually, I think we should allow for grid values lower or larger than the grid, since we allow for methods that support extrapolation, such as `fill_extend` option.

Reply:
okay, this makes sense!